legate/util/settings.py

```diff
@@ -0,0 +1,307 @@
+
+    def __call__(
+        self, value: T | str | None = None, default: Unset[T] = _Unset
+    ) -> T:
+        """Return the setting value according to the standard precedence.
+
+        Args:
+            value (any, optional):
+                An optional immediate value. If not None, the value will
+                be converted, then returned.
+
+            default (any, optional):
+                An optional default value that only takes precendence over
+                implicit default values specified on the property itself.
+
+        Returns:
+            str or int or float
+
+        Raises:
+            RuntimeError
+        """
+
+        # 4. immediate values
+        if value is not None:
+            return self._convert(value)
+
+        # 3. previously user-set value
+        if self._user_value is not _Unset:
+            return self._convert(self._user_value)
+
+        # 2. environment variable
+        if self._env_var and self._env_var in os.environ:
+            return self._convert(os.environ[self._env_var])
+
+        # 1. local defaults
+        if default is not _Unset:
+            return self._convert(default)
+
+        # 0. global defaults
+        if self._default is not _Unset:
+            return self._convert(self._default)
+
+        raise RuntimeError(
+            f"No configured value found for setting {self._name!r}"
+        )
+
+    def __get__(
+        self, instance: Any, owner: type[Any]
+    ) -> PrioritizedSetting[T]:
+        return self
+
+    def __set__(self, instance: Any, value: str | T) -> None:
+        self.set_value(value)
+
+    def set_value(self, value: str | T) -> None:
+        """Specify a value for this setting programmatically.
+
+        A value set this way takes precedence over all other methods except
+        immediate values.
+
+        Args:
+            value (str or int or float):
+                A user-set value for this setting
+
+        Returns:
+            None
+        """
+        # It is usually not advised to store any data directly on descriptors,
+        # since they are shared by all instances. But in our case we only ever
+        # have a single instance of a given settings object.
+        self._user_value = value
+
+    def unset_value(self) -> None:
+        """Unset the previous user value such that the priority is reset."""
+        self._user_value = _Unset
+
+    @property
+    def env_var(self) -> str | None:
+        return self._env_var
+
+    @property
+    def default(self) -> Unset[T]:
+        return self._default
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def help(self) -> str:
+        return self._help
+
+    @property
+    def convert_type(self) -> str:
+        if self._convert is convert_str:
+            return "String"
+        if self._convert is convert_bool:
+            return "Bool"
+        if self._convert is convert_str_seq:
+            return "List[String]"
```

Comment:
It looks like `convert_str_seq` returns `tuple[str, ...]`.

Reply:
f83dfba